### PR TITLE
mgr/stats: add manager module for performance stats

### PR DESCRIPTION
**Pull request by vshankar**
This is unusable right now since it needs pr #26004 and a set of other patches that add mgr interfaces to add, remove user query and get performance counters from ceph-mds.

command line:
`$ ceph fs perf stats {<mds_rank>} :  retrieve ceph fs performance stats`

Note that the above command does not continuously display perf stats -- it just shows perf metrics available to ceph-mgr at the time of invocation. There's needs to be a wrapper interface (something like `ceph fs top`) that invokes this command, consumes the output json and displays similar to `top`.

Some sample invocations (via `watch`):

`ceph fs perf stats` with `max_mds=1`, 3 clients (show metrics from all MDSs)

ceph-fs-perf-stat-1

`ceph fs perf stats` with `max_mds=2`, 2 clients (show metrics from all MDSs)

ceph-fs-perf-stat-2

`ceph fs perf stats --mds_rank=1` with `max_mds=2`, 2 clients (show metrics from MDS rank 1)

ceph-fs-perf-stat-3

NOTE: only `cap_hit_metric` (as `hits,misses`) are shown above.

### CODE REVIEW

**Comment by vshankar** [comment]
remove draft/WIP state -- rebased on top on pr #26004 (contains commits 3464a45410..f5488d1666)

**Comment by lxbsz** [comment]
@vshankar  BTW, should be give some hints about what the following numbers mean in the output ? For the other users they may need to figure it out from the source code, maybe we can fix it in the "counters" and "global_counters" sections.
```
            [
                31909,
                563000000
            ]
```

The output:

```
{
    "client_metadata": {
        "client.7772": {
            "IP": "v1:x.x.x.x"
        },
        "client.7814": {
            "IP": "x.x.x.x",
            "hostname": "mynode",
            "mount_point": "N/A",
            "root": "/"
        }
    },
    "counters": [
        "cap_hit"
    ],
    "global_counters": [
        "read_latency",
        "write_latency",
        "metadata_latency"
    ],
    "global_metrics": {
        "client.7772": [
            [
                0,
                0
            ],
            [
                0,
                0
            ],
            [
                31909,
                563000000
            ]
        ]
    },
    "metrics": {
        "delayed_ranks": [],
        "mds.0": {
            "client.7772": [
                [
                    0,
                    0
                ]
            ]
        }
    }
}
```

**Comment by vshankar** [comment]
> @vshankar BTW, should be give some hints about what the following numbers mean in the output ? For the others they may need to figure them out from the source code, maybe we can fix it in the "counters" and "global_counters" sections.

good idea -- but I'm a bit wary on including descriptions in the mgr command output itself. We could precisely document the counters and what each one signifies?

**Comment by lxbsz** [comment]
> > @vshankar BTW, should be give some hints about what the following numbers mean in the output ? For the others they may need to figure them out from the source code, maybe we can fix it in the "counters" and "global_counters" sections.
> 
> good idea -- but I'm a bit wary on including descriptions in the mgr command output itself. We could precisely document the counters and what each one signifies?

This maybe a good alternative.

**Comment by lxbsz** [comment]
This looks good to me.

@vshankar @batrick 
BTW, both https://tracker.ceph.com/issues/46865 and https://github.com/ceph/ceph/pull/32479 will depend on this PR.

Thanks

**Comment by vshankar** [comment]
> This looks good to me.

Thanks @lxbsz. I'll carve out one of the commits (8286e0036a36a3748bb2b2aaa0787599ab2462d5) as a separate PR so that it can be merged before this...

**Comment by lxbsz** [comment]
> > This looks good to me.
> 
> Thanks @lxbsz. I'll carve out one of the commits ([8286e00](https://github.com/ceph/ceph/commit/8286e0036a36a3748bb2b2aaa0787599ab2462d5)) as a separate PR so that it can be merged before this...

Okay, that commit looks good to me too.

**Comment by batrick** [comment]
@vshankar this PR is ready for review?

**Comment by batrick** [comment]
jenkins test make check

**Comment by vshankar** [comment]
> @vshankar this PR is ready for review?

One commit can be pushed as a separate PR. I'll ping when ready.

**Comment by vshankar** [comment]
> @vshankar this PR is ready for review?

oh yeh! I need to add tox tests for this (manager modules in general).

**Comment by vshankar** [comment]
> Okay, that commit looks good to me too.

This commit is no longer needed for this pr. I have removed it w/ latest update.

**Comment by vshankar** [comment]
@batrick please take look.

**Comment by vshankar** [comment]
jenkins test make check

File: qa/tasks/cephfs/filesystem.py [hunk 1]
```diff
@@ -1440,3 +1489,44 @@ def data_scan(self, args, quiet=False, worker_count=1):
 
     def is_full(self):
         return self.is_pool_full(self.get_data_pool_name())
+
+    def grow(self, new_max_mds):
```
Comment (batrick) [inline qa/tasks/cephfs/filesystem.py:1493]: In other tests we don't really go to all of this trouble. Instead just do:

````
self.fs.set_max_mds(foo)
status = self.fs.wait_for_daemons()
````

There's not much reason to keep any of these extra checks.

File: qa/tasks/cephfs/filesystem.py [hunk 2]
```diff
@@ -1440,3 +1489,44 @@ def data_scan(self, args, quiet=False, worker_count=1):
...
+
+        grace = float(self.get_config("mds_beacon_grace", service_type="mon"))
+
+        log.info("Waiting for cluster to grow.")
+        fs_status = self.wait_for_daemons(timeout=60+grace*2)
+        ranks = set([info['gid'] for info in fs_status.get_ranks(fscid)])
+        assert(original_ranks.issubset(ranks) and len(ranks) == new_max_mds)
+        return fs_status
+
+    def shrink(self, new_max_mds):
```
Comment (batrick) [inline qa/tasks/cephfs/filesystem.py:1513]: ditto

File: qa/tasks/cephfs/filesystem.py [hunk 3]
```diff
@@ -410,6 +421,44 @@ def set_block(_mds_id):
 
         self._one_or_all(mds_id, set_block, in_parallel=False)
 
+    def set_inter_mds_block(self, blocked, mds_id1, mds_id2):
```
[the remark this answers is not in the document]
Comment (vshankar) [reply qa/tasks/cephfs/filesystem.py:424]: ACK (I kind of followed set_clients_block() that works on mds names)

**Comment by vshankar** [comment]
> I'm still going through the rest. Will post another review later.

Hey Patrick -- are you still going through this PR? I have fixed your comments, but waiting for more reviews...

**Comment by batrick** [comment]
> > I'm still going through the rest. Will post another review later.
> 
> Hey Patrick -- are you still going through this PR? I have fixed your comments, but waiting for more reviews...

Unless I'm mistaken, I think you haven't pushed the changes yet so I was waiting for the update. Sorry I should have said so. In any case, this needs rebased.

**Comment by vshankar** [comment]
> Unless I'm mistaken, I think you haven't pushed the changes yet so I was waiting for the update. Sorry I should have said so. In any case, this needs rebased.

no worries -- I'll rebase and update.

**Comment by vshankar** [comment]
@batrick rebased and updated. please take a look.

**Comment by batrick** [comment]
retest this please

File: src/mgr/ActivePyModules.cc [hunk 4]
```diff
@@ -1054,16 +1054,80 @@ void ActivePyModules::remove_osd_perf_query(MetricQueryID query_id)
...
     Py_RETURN_NONE;
   }
 
   PyFormatter f;
+  const std::map<OSDPerfMetricKey, PerformanceCounters> &counters = collector.counters;
+
+  f.open_array_section("counters");
+  for (auto &it : counters) {
+    auto &key = it.first;
+    auto  &instance_counters = it.second;
```
Comment (batrick) [inline src/mgr/ActivePyModules.cc:1071]:
```suggestion
  for (auto& [key, instance_counters] : counters) {
```

File: src/mgr/BaseMgrModule.cc [hunk 5]
```diff
@@ -1023,6 +1023,261 @@ ceph_get_osd_perf_counters(BaseMgrModule *self, PyObject *args)
...
+  //   ],
+  //   'limit': {'order_by': performance_counter_type, 'max_count': n},
+  // }
+
+  for (int i = 0; i < PyList_Size(query_params); ++i) {
+    PyObject *kv = PyList_GET_ITEM(query_params, i);
+    char *query_param_name = nullptr;
+    PyObject *query_param_val = nullptr;
+    if (!PyArg_ParseTuple(kv, "sO:pair", &query_param_name, &query_param_val)) {
+      derr << __func__ << " dict item " << i << " not a size 2 tuple" << dendl;
```
Comment (batrick) [inline src/mgr/BaseMgrModule.cc:1080]: I'd prefer we throw an exception for these. What do you think?

File: src/mgr/BaseMgrModule.cc [hunk 6]
```diff
@@ -1023,6 +1023,261 @@ ceph_get_osd_perf_counters(BaseMgrModule *self, PyObject *args)
...
+      Py_RETURN_NONE;
+    }
+    if (query_param_name == NAME_KEY_DESCRIPTOR) {
```
Comment (batrick) [inline src/mgr/BaseMgrModule.cc:1083]: This method is too long and hard to follow. Can you split it up and add comments please.

**Review by batrick — CHANGES_REQUESTED** [review]
Module is hard for me to follow what is going on, probably because we haven't discussed this in a while. I'll arrange a meeting after standup so we can discuss.

There also needs to be documentation for the `fs perf stats` command.

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+                return False
+        global_metrics = metrics['global_metrics']
+        client_metadata = metrics['client_metadata']
+        for client in list(client_metadata.keys()):
+            if client_metadata[client]['mount_point'] == "N/A":
+                del client_metadata[client]
+                del global_metrics[client]
+        if not len(global_metrics) == client_count or not len(client_metadata) == client_count:
+            return False
+        return True
```
[the remark this answers is not in the document]
Comment (vshankar) [reply qa/tasks/cephfs/test_mds_metrics.py:295]: ack -- will fix.

**Comment by vshankar** [comment]
https://pulpito.ceph.com/vshankar-2020-10-01_09:36:25-multimds-wip-fs-perf-stats-20201001-041335-distro-basic-smithi/

Will update this PR with the changes and ping when ready for review.

**Comment by vshankar** [comment]
Run link: https://pulpito.ceph.com/vshankar-2020-10-06_06:16:13-multimds-wip-fs-perf-stats-20201006-001506-distro-basic-smithi/

The failed tests are due to absence of `mount_point` entry client metadata (fetched by the stats module).

```
{'id': 4946, 'entity': {'name': {'type': 'client', 'num': 4946}, 'addr': {'type': 'v1', 'addr': '172.21.15.71:0', 'nonce': 278072023}}, 'state': 'open', 'num_leases': 0, 'num_caps': 1, 'request_load_avg': 0, 'uptime': 9.889729723, 'requests_in_flight': 0,'completed_requests': [], 'reconnecting': False, 'recall_caps': {'value': 0, 'halflife': 60}, 'release_caps': {'value': 0, 'halflife': 60}, 'recall_caps_throttle': {'value': 0, 'halflife': 2.5}, 'recall_caps_throttle2o': {'value': 0, 'halflife': 0.5}, 'session_cache_liveness': {'value': 0.9774134759196786, 'halflife': 300}, 'inst': 'client.4946 v1:172.21.15.71:0/278072023', 'prealloc_inos': [], 'used_inos': [], 'client_metadata': {'client_features': {'feature_bits': '0x0000000000007bff'}, 'metric_spec': {'metric_flags' {'feature_bits':'0x000000000000001f'}}, 'entity_id': '1', 'hostname': 'smithi071', 'kernel_version': '5.9.0-rc4-ceph-ga8ae7497c333', 'root': '/'}}
```

This is a kclient with kernel version: 5.9.0-rc4-ceph-ga8ae7497c333. Why would the `mount_point` metadata be missing?

**Comment by vshankar** [comment]
> Run link: https://pulpito.ceph.com/vshankar-2020-10-06_06:16:13-multimds-wip-fs-perf-stats-20201006-001506-distro-basic-smithi/
> 
> The failed tests are due to absence of `mount_point` entry client metadata (fetched by the stats module).
> 
> ```
> {'id': 4946, 'entity': {'name': {'type': 'client', 'num': 4946}, 'addr': {'type': 'v1', 'addr': '172.21.15.71:0', 'nonce': 278072023}}, 'state': 'open', 'num_leases': 0, 'num_caps': 1, 'request_load_avg': 0, 'uptime': 9.889729723, 'requests_in_flight': 0,'completed_requests': [], 'reconnecting': False, 'recall_caps': {'value': 0, 'halflife': 60}, 'release_caps': {'value': 0, 'halflife': 60}, 'recall_caps_throttle': {'value': 0, 'halflife': 2.5}, 'recall_caps_throttle2o': {'value': 0, 'halflife': 0.5}, 'session_cache_liveness': {'value': 0.9774134759196786, 'halflife': 300}, 'inst': 'client.4946 v1:172.21.15.71:0/278072023', 'prealloc_inos': [], 'used_inos': [], 'client_metadata': {'client_features': {'feature_bits': '0x0000000000007bff'}, 'metric_spec': {'metric_flags' {'feature_bits':'0x000000000000001f'}}, 'entity_id': '1', 'hostname': 'smithi071', 'kernel_version': '5.9.0-rc4-ceph-ga8ae7497c333', 'root': '/'}}
> ```
> 
> This is a kclient with kernel version: 5.9.0-rc4-ceph-ga8ae7497c333. Why would the `mount_point` metadata be missing?

BTW, client metadata is fetched by doing an asynchronous query ("client ls") to an MDS.

**Comment by vshankar** [comment]
> AFAIK, `mount_point` has never been shared by the kernel driver. You can see this information in `mds_client.c` in `create_session_open_msg`.

Ummm.. I recall seeing `mount_point` metadata with kclient -- maybe I was running a new(er) kernel.

**Comment by vshankar** [comment]
@batrick please take a look. Tests pass: https://pulpito.ceph.com/vshankar-2020-10-08_11:02:25-multimds-wip-fs-perf-stats-20201008-054432-distro-basic-smithi/

Should we also run rados/mgr tests too?

File: src/mgr/BaseMgrModule.cc (hunk 5 above)
Comment (vshankar) [reply src/mgr/BaseMgrModule.cc:1081]: follows osd metrics template

File: src/mgr/BaseMgrModule.cc (hunk 6 above)
Comment (vshankar) [reply src/mgr/BaseMgrModule.cc:1084]: ditto -- follows osd metrics template. I have mentioned that in the routine.

File: qa/tasks/cephfs/test_mds_metrics.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (vshankar) [reply qa/tasks/cephfs/test_mds_metrics.py]: done

**Comment by vshankar** [comment]
retest this please

**Comment by vshankar** [comment]
@batrick -- regarding the documentation for this, I will include that with the PR that adds the `fs top` utility.

**Comment by batrick** [comment]
> @batrick please take a look. Tests pass: https://pulpito.ceph.com/vshankar-2020-10-08_11:02:25-multimds-wip-fs-perf-stats-20201008-054432-distro-basic-smithi/
> 
> Should we also run rados/mgr tests too?

Yes, I think so.

**Comment by batrick** [comment]
jenkins test make check

**Comment by vshankar** [comment]
rados/mgr tests: https://pulpito.ceph.com/vshankar-2020-10-08_16:46:20-rados:mgr-wip-fs-perf-stats-20201008-054432-distro-basic-smithi/

**Review by batrick — CHANGES_REQUESTED** [review]
````
2020-10-09T20:05:23.296 INFO:tasks.cephfs_test_runner:======================================================================
2020-10-09T20:05:23.296 INFO:tasks.cephfs_test_runner:ERROR: test_up_less_than_max (tasks.cephfs.test_failover.TestClusterResize)
2020-10-09T20:05:23.297 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-10-09T20:05:23.297 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-10-09T20:05:23.297 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20201009.020122/qa/tasks/cephfs/test_failover.py", line 204, in test_up_less_than_max
2020-10-09T20:05:23.297 INFO:tasks.cephfs_test_runner:    self.shrink(2)
2020-10-09T20:05:23.298 INFO:tasks.cephfs_test_runner:AttributeError: 'TestClusterResize' object has no attribute 'shrink'
````

From: https://pulpito.ceph.com/pdonnell-2020-10-09_05:07:33-fs-wip-pdonnell-testing-20201009.020122-distro-basic-smithi/5508603/

After fixing that please rerun those tests and then I think it's good to merge!

**Comment by vshankar** [comment]
@batrick fixed and updated

test run for failed tests: https://pulpito.ceph.com/vshankar-2020-10-12_08:41:52-fs-wip-fs-perf-stats-20201012-023913-distro-basic-smithi/

**Comment by batrick** [comment]
jenkins test make check

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
\o/

**Comment by lxbsz** [comment]
Cool, I will begin to work on my PRs which depends on this one. Thanks.
